### PR TITLE
core/txpool/legacypool: respect nolocals-setting

### PR DESCRIPTION
**Pull request by jp-imx**
I've added a check for `pool.config.NoLocals` when setting up `txpool` validation. The goal is to reject transactions if they don't meet the `--txpool.pricelimit` setting.

### CODE REVIEW

File: core/txpool/legacypool/legacypool.go [hunk 1]
```diff
@@ -585,7 +585,7 @@ func (pool *LegacyPool) validateTxBasics(tx *types.Transaction, local bool) erro
 		MaxSize: txMaxSize,
 		MinTip:  pool.gasTip.Load(),
 	}
-	if local {
+	if local && !pool.config.NoLocals {
```
Comment (lightclient) [inline core/txpool/legacypool/legacypool.go:588]: Negating a negative is not the most readable. I don't have a great solution, although two ideas:
* rename `NoLocals` to `PrioritizeLocals` or something of that sort
* add a helper func on `config` with a name like `PrioritizeLocals` so the negating a negative happens behind the scenes and the rest of the logic is very readable.

**Review by lightclient — COMMENTED** [review]
Looks good, one minor nit.

File: core/txpool/legacypool/legacypool.go (hunk 1 above)
Comment (jp-imx) [reply core/txpool/legacypool/legacypool.go]: @lightclient addressed with the `PrioritizeLocals` wrapper, as it seemed like the simplest and least invasive approach

**Review by MariusVanDerWijden — CHANGES_REQUESTED** [review]
Looks like during the big refactor, this code got broken a bit.
Before that, local transactions were enqueued using `addLocals` which sets `local` to `!pool.config.NoLocals`. Since it is a subpool now, the transactions are added using `Add` with `local` set to true if they originated by the rpc. Thus we never check if `noLocals` is set. This means that we allow locals _even if_ `noLocals` is set to false. 

I think the correct fix would be something like this:
```
if local && pool.config.NoLocals {
    local = false
}
```

**Comment by jp-imx** [comment]
@MariusVanDerWijden, thanks for that insight. I've reshaped my changes to match your suggestion. I've also updated my tests to make sure I cover `addLocal` and `Add`.

**Comment by jp-imx** [comment]
@MariusVanDerWijden actually, I had to revert that approach. The `local` flag doesn't propagate to the validation logic. So we were rejecting transactions even when the `NoLocals` flag was set to `false`.

**Comment by jp-imx** [comment]
Hi @MariusVanDerWijden @lightclient @holiman @rjl493456442 @karalabe 

I just wanted to follow up on this PR.

Cheers!

File: core/txpool/legacypool/legacypool.go [hunk 2]
```diff
@@ -585,7 +589,7 @@ func (pool *LegacyPool) validateTxBasics(tx *types.Transaction, local bool) erro
 		MaxSize: txMaxSize,
 		MinTip:  pool.gasTip.Load(),
 	}
-	if local {
+	if local && pool.config.PrioritizeLocals() {
```
Comment (holiman) [inline core/txpool/legacypool/legacypool.go:592]: as far as I can tell, `local` will never be if `config.NoLocals` is set.  In your testcase, you call 

```
	if err := pool.addLocal(tx); !errors.Is(err, txpool.ErrUnderpriced) {
		t.Fatalf("Min tip not enforced")
	}

```
which invokes
```
func (pool *LegacyPool) addLocal(tx *types.Transaction) error {
	errs := pool.addLocals([]*types.Transaction{tx})
	return errs[0]
}

```
-> 
```
func (pool *LegacyPool) addLocals(txs []*types.Transaction) []error {
	return pool.addTxs(txs, !pool.config.NoLocals, true)
}
```
i.e `addTxs( txs, false, true ) `. So `local` is `false`. 

The test passes even without this change. Not sure what the original problem were that you tried to solve.

**Review by holiman — DISMISSED** [review]

File: core/txpool/legacypool/legacypool.go (hunk 2 above)
Comment (jp-imx) [reply core/txpool/legacypool/legacypool.go:592]: @holiman, thanks for picking that up. The tests were missing checks. I lost them while reverting some changes suggested during the review that didn't pan out.

The issue I'm addressing is not with `addLocal`. The issue I'm addressing is with how local transactions are added using `Add` with local set to `true` after "the big refactor" (see @MariusVanDerWijden 's analysis  [here](https://github.com/ethereum/go-ethereum/pull/28435#pullrequestreview-1705640245)). 

tl;dr: right now, even if we set `txpool.nolocals`, `validateTxBasics` always sets `opts.MinTip` to `0` and `txpool.pricelimit` is not enforced. My PR fixes that.

**Comment by jp-imx** [comment]
@rjl493456442 it's where I branched off when I started my work. I can re-target. TBH, it shouldn't be a big issue. Is that the preference then?

File: core/txpool/legacypool/legacypool.go (hunk 2 above)
Comment (holiman) [reply core/txpool/legacypool/legacypool.go:592]: I see. I think I'd prefer doing the fix earlier: 
```diff
diff --git a/core/txpool/legacypool/legacypool.go b/core/txpool/legacypool/legacypool.go
index 0e33923274..853121421c 100644
--- a/core/txpool/legacypool/legacypool.go
+++ b/core/txpool/legacypool/legacypool.go
@@ -959,6 +959,7 @@ func (pool *LegacyPool) addRemoteSync(tx *types.Transaction) error {
 // If sync is set, the method will block until all internal maintenance related
 // to the add is finished. Only use this during tests for determinism!
 func (pool *LegacyPool) Add(txs []*types.Transaction, local, sync bool) []error {
+	local = local && !pool.config.NoLocals
 	// Filter out known ones without obtaining the pool lock or recovering signatures
 	var (
 		errs = make([]error, len(txs))

```

Comment (jp-imx) [reply core/txpool/legacypool/legacypool.go]: @holiman thanks for the feedback. I've applied your suggestion

**Comment by jp-imx** [comment]
@rjl493456442 I've rebased my changes and updated the target branch

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

**Comment by MariusVanDerWijden** [comment]
Closing and reopening for ci

**Comment by jp-imx** [comment]
Hey @MariusVanDerWijden @rjl493456442 @holiman, is this ready to merge? Just pointing out that I don't have permission to merge PRs into the repository.

**Review by holiman — APPROVED** [review]
LGTM
